filter.go

```diff
@@ -20,6 +20,11 @@ func SubtractFilter(addrs ...ma.Multiaddr) func(ma.Multiaddr) bool {
 // IsFDCostlyTransport returns true for transports that require a new file
 // descriptor per connection created
 func IsFDCostlyTransport(a ma.Multiaddr) bool {
+	mas := ma.Split(a)
+	if len(mas) < 2 {
+		return false
+	}
+	a = mas[0].Encapsulate(mas[1])
```

Comment:
this seems pretty memory expensive...

Reply:
better idea how to do it?

Reply:
A single `a.Protocols()` and then a check whether any of them is `P_TCP`, should simplify this a lot. Then `mafmt.TCP.matches(a)` can be dropped, which actually does quite a lot under the hood. Collecting only the protocols is less work than building up the whole thing and then matching. (We should amend multiaddr-fmt to the same, but that's for another time.)

That would also make this function more future-proof, TCP won't neccessarily always be at position 2 in the address, but it'll always be FD-costly.